yarn-project/simulator/src/avm/avm_tree.ts

```diff
@@ -0,0 +1,808 @@
+        newOrElementToUpdate: { update: true, element: existingPublicData },
+        lowWitness: lowResult,
+      };
+    }
+    // We are writing to a new slot
+    const insertionIndex = tree.leafCount;
+    const lowWitness = lowResult.preimage;
+
+    const updateLowWitness = new PublicDataTreeLeafPreimage(
+      lowWitness.slot,
+      lowWitness.value,
+      new Fr(slot.toBigInt()),
+      insertionIndex,
+    );
+
+    const newPublicDataNode = new PublicDataTreeLeafPreimage(
+      slot,
+      newValue,
+      new Fr(lowWitness.getNextKey()),
+      lowWitness.getNextIndex(),
+    );
+    const insertionPath = await this.appendIndexedTree(treeId, lowResult.index, updateLowWitness, newPublicDataNode);
+
+    // Since we are appending, we might have a new minimum public data leaf
+    this._updateMinInfo(
+      MerkleTreeId.PUBLIC_DATA_TREE,
+      PublicDataTreeLeafPreimage,
+      [newPublicDataNode, updateLowWitness],
+      [insertionIndex, lowResult.index],
+    );
+    return {
+      leafIndex: insertionIndex,
+      insertionPath: insertionPath,
+      newOrElementToUpdate: { update: false, element: newPublicDataNode },
+      lowWitness: lowResult,
+    };
+  }
+
+  /**
+   * This is just a helper to compare the preimages and update the minimum public data leaf
+   * @param treeId - The tree to be queried for a sibling path.
+   * @param T - The type of the preimage (PublicData or Nullifier)
+   * @param preimages - The preimages to be compared
+   * @param indices - The indices of the preimages
+   */
+  private _updateMinInfo<T extends IndexedTreeLeafPreimage>(
+    treeId: IndexedTreeId,
+    T: { fromBuffer: (buffer: Buffer) => T },
+    preimages: T[],
+    indices: bigint[],
+  ): void {
+    let currentMin = this.getMinInfo(treeId, T);
+    if (currentMin === undefined) {
+      currentMin = { preimage: preimages[0], index: indices[0] };
+    }
+    for (let i = 0; i < preimages.length; i++) {
+      if (preimages[i].getKey() <= currentMin.preimage.getKey()) {
+        currentMin = { preimage: preimages[i], index: indices[i] };
+      }
+    }
+    this.setMinInfo(treeId, currentMin.preimage, currentMin.index);
+  }
+
+  /**
+   * This appends a nullifier to the nullifier tree, and throws if the nullifier already exists
+   * @param value - The nullifier to be appended
+   * @returns The insertion result which contains the insertion path, low leaf and the new leaf index
+   */
+  async appendNullifier(value: Fr): Promise<IndexedInsertionResult<NullifierLeafPreimage>> {
+    const treeId = MerkleTreeId.NULLIFIER_TREE;
+    const tree = this.treeMap.get(treeId)!;
+    const lowResult = await this._getIndexedLowWitness(treeId, value, NullifierLeafPreimage);
+    if (lowResult.update) {
+      throw new Error('Not allowed to update nullifier');
+    }
+    // We are writing a new entry
+    const insertionIndex = tree.leafCount;
+    const lowWitness = lowResult.preimage;
```

Comment:
The naming is confusing here. I think if you stick to what you've done in this file, lowResult should become lowWitness and lowWitness should become lowPreimage. But I'm not sure I like using "witness" for this....